pkg/kubelet/kubelet_pods.go

```diff
@@ -1112,70 +1133,45 @@ func (kl *Kubelet) HandlePodCleanups(ctx context.Context) error {
 	// that have already been removed from config. Pods that are terminating
 	// will be added to possiblyRunningPods, to prevent overly aggressive
 	// cleanup of pod cgroups.
+	stringIfTrue := func(t bool) string {
+		if t {
+			return "true"
+		}
+		return ""
+	}
 	runningPods := make(map[types.UID]sets.Empty)
 	possiblyRunningPods := make(map[types.UID]sets.Empty)
-	restartablePods := make(map[types.UID]sets.Empty)
 	for uid, sync := range workingPods {
-		switch sync {
+		switch sync.State {
 		case SyncPod:
 			runningPods[uid] = struct{}{}
 			possiblyRunningPods[uid] = struct{}{}
 		case TerminatingPod:
 			possiblyRunningPods[uid] = struct{}{}
-		case TerminatedAndRecreatedPod:
-			restartablePods[uid] = struct{}{}
+		default:
 		}
 	}
 
-	// Stop probing pods that are not running
-	klog.V(3).InfoS("Clean up probes for terminated pods")
-	kl.probeManager.CleanupPods(possiblyRunningPods)
-
-	// Terminate any pods that are observed in the runtime but not
-	// present in the list of known running pods from config.
+	// Retrieve the list of running containers from the runtime to perform cleanup.
+	// We need the latest state to avoid delaying restarts of static pods that reuse
+	// a UID.
+	if err := kl.runtimeCache.ForceUpdateIfOlder(ctx, kl.clock.Now()); err != nil {
```

Comment:
Yes, we did.  And it caused a race that resulted in the pod getting stuck another way (because of another bug in my PR that I later fixed).  Roughly because the cache was older you might have to wait an extra 2s to restart a static pod.  Only a few callers use the runtime cache, but this is the one place (resync) where we don't want to reuse the cache. I debated changing this code to be the *only* refresher of the cache, but that was wierd too.  Ideally the cache interval is slightly longer than the cleanups loop (because HandlePodCleanups needs to be really fast) so we always refresh the cache here but never revalidate it, but that isn't happening right now.

Reply:
I aligned these in the kubelet (see NewRuntimeCache())

Reply:
I also tightened the housekeeping warning duration to 1s from 15s, and make the cache expiration interval = housekeeping interval + housekeeping warning interval.  If housekeeping is taking 15s something bad is happening - we might get more errors from it but nothing in there should take 15s.

Reply:
got it, thanks for the explanation and makes sense re cache interval updates.